packages/labs/virtualizer/src/test/scenarios/fixed-position-ancestor.test.ts

```diff
@@ -0,0 +1,119 @@
+    // space, making our testing job easier.
+    const container = await fixture(html`
+      <div id="container" style=${styleMap(containerStyles)}>
+        <div id="scroller" style=${styleMap(scrollerStyles)}>
+          <lit-virtualizer
+            .layout=${grid({
+              itemSize: '100px',
+            })}
+            .items=${_100Items}
+            .renderItem=${({name}: NamedItem) => html`<div>${name}</div>`}
+          ></lit-virtualizer>
+        </div>
+      </div>
+    `);
+
+    const scroller = container.querySelector('#scroller')!;
+    expect(scroller).to.be.instanceOf(HTMLDivElement);
+    const virtualizer = container.querySelector('lit-virtualizer')!;
+    expect(virtualizer).to.be.instanceOf(LitVirtualizer);
+
+    // If virtualizer has properly ignored the zero-width ancestor of our
+    // fixed-position scroller, some children will be rendered; otherwise, not.
+    //
+    // In practice, we'll time out if we fail here because the `layoutComplete`
+    // promise will never be fulfilled.
+    await virtualizer.layoutComplete;
+    expect(virtualizer.textContent).to.contain('Item 0');
+  });
+
+  // Regression test for https://github.com/lit/lit/issues/4125
+  it('should respond to scroll events from a fixed-position scroller', async () => {
+    const containerStyles = {
+      overflow: 'hidden',
+    };
+
+    // Our scroller is position: fixed. While its ancestors (if any) should
+    // therefore be ignored for the purpose of calculating virtualizer's
+    // viewport, the scroller itself should be considered a clipping ancestor,
+    // and virtualizer should listen to any `scroll` events it emits.
+    const scrollerStyles = {
+      position: 'fixed',
+      width: '200px',
+      height: '200px',
+      overflow: 'auto',
+    };
+
+    const container = await fixture(html`
+      <div id="container" style=${styleMap(containerStyles)}>
+        <div id="scroller" style=${styleMap(scrollerStyles)}>
+          <lit-virtualizer
+            .items=${_100Items}
+            .renderItem=${({name}: NamedItem) => html`<p>${name}</p>`}
+          ></lit-virtualizer>
+        </div>
+      </div>
+    `);
+
+    const scroller = container.querySelector('#scroller')!;
+    expect(scroller).to.be.instanceOf(HTMLDivElement);
+    const virtualizer = container.querySelector('lit-virtualizer')!;
+    expect(virtualizer).to.be.instanceOf(LitVirtualizer);
+
+    await virtualizer.layoutComplete;
+    expect(virtualizer.textContent).to.contain('Item 0');
+
+    // If the position: fixed scroller has properly been recognized as
+    // a clipping ancestor, then virtualizer will re-render as scrolling
+    // occurs; otherwise, not.
+    //
+    // In practice, we'll time out if we fail here because the `layoutComplete`
+    // promise will never be fulfilled.
+    scroller.scrollTo(0, scroller.scrollHeight);
+    await virtualizer.layoutComplete;
+    expect(virtualizer.textContent).to.contain('Item 99');
```

Comment:
Interestingly replacing both of the `await virtualizer.layoutComplete` here with `pass()` makes this fail. There must be something about starting a `scroller.scrollTo()` specifically after awaiting `layoutComplete` that looking for `'Item 0'` alone isn't enough.

Reply:
Yeah, it is possible in a testing context to initiate a scroll while virtualizer is still in the middle of a layout pass, and this can cause unpredictable results. Waiting for `layoutComplete` avoids this situation, which is another reason why I'd like to lean on it more heavily in our tests.

In my updated version of this test, I added some ad hoc logic to race the `layoutComplete` promise against a short timeout so that we don't fail with a timeout; instead, we fail because the rendered state of the virtualizer is not correct (which aligns nicely with how you'd observe this failure in manual testing).